pkg/manager/member/tidb_init_manager.go

```diff
@@ -0,0 +1,428 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package member
+
+import (
+	"fmt"
+	"path"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	batchlisters "k8s.io/client-go/listers/batch/v1"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/client/clientset/versioned"
+	listers "github.com/pingcap/tidb-operator/pkg/client/listers/pingcap/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/controller"
+	"github.com/pingcap/tidb-operator/pkg/label"
+)
+
+const (
+	initContainerName   = "wait"
+	containerName       = "mysql-client"
+	passwdKey           = "password"
+	passwdPath          = "/etc/tidb/password"
+	sqlKey              = "init-sql"
+	sqlPath             = "init.sql"
+	sqlDir              = "/data"
+	startScriptPath     = "start_script.py"
+	initStartScriptPath = "init_start_script.sh"
+	startScriptDir      = "/usr/local/bin"
+	startKey            = "start-script"
+	initStartKey        = "init-start-script"
+	// initContainerImage is the image for init container
+	initContainerImage = "busybox:1.31.1"
+)
+
+// InitManager implements the logic for syncing TidbInitializer.
+type InitManager interface {
+	// Sync	implements the logic for syncing TidbInitializer.
+	Sync(*v1alpha1.TidbInitializer) error
+}
+
+type tidbInitManager struct {
+	jobLister    batchlisters.JobLister
+	cli          versioned.Interface
+	tiLister     listers.TidbInitializerLister
+	typedControl controller.TypedControlInterface
+}
+
+// NewTiDBInitManager return tidbInitManager
+func NewTiDBInitManager(
+	jobLister batchlisters.JobLister,
+	cli versioned.Interface,
+	tiLister listers.TidbInitializerLister,
+	typedControl controller.TypedControlInterface,
+) InitManager {
+	return &tidbInitManager{
+		jobLister,
+		cli,
+		tiLister,
+		typedControl,
+	}
+}
+
+func (tm *tidbInitManager) Sync(ti *v1alpha1.TidbInitializer) error {
+	name := controller.TiDBInitializerMemberName(ti.Spec.Clusters.Name)
+	ns := ti.Namespace
+	cm := &corev1.ConfigMap{}
+
+	exist, err := tm.typedControl.Exist(client.ObjectKey{
+		Namespace: ns,
+		Name:      name,
+	}, cm)
+	if err != nil {
+		return err
+	}
+	if !exist {
+		err = tm.syncTiDBInitConfigMap(ti)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
For a one-time deal, I agree that we can just create the object, but here is the logic in sync, it will be done at least periodically, it's not reasonable to create the object each time.

Reply:
Check before creation can reduce the API call to kube-apiserver, and I have moved the check block to separate sync functions.